// magic-sdk/dist/cjs/index.js
"use strict";
/* eslint-disable no-underscore-dangle */
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var provider_1 = require("@magic-sdk/provider");
var localforage_1 = tslib_1.__importDefault(require("localforage"));
var memoryDriver = tslib_1.__importStar(require("localforage-driver-memory"));
var iframe_controller_1 = require("./iframe-controller");
var web_transport_1 = require("./web-transport");
tslib_1.__exportStar(require("@magic-sdk/commons"), exports);
exports.Magic = provider_1.createSDK(provider_1.SDKBase, {
    platform: 'web',
    sdkName: 'magic-sdk',
    version: '5.1.0',
    defaultEndpoint: 'https://auth.magic.link/',
    ViewController: iframe_controller_1.IframeController,
    PayloadTransport: web_transport_1.WebTransport,
    configureStorage: /* istanbul ignore next */ function () { return tslib_1.__awaiter(void 0, void 0, void 0, function () {
        var lf;
        return tslib_1.__generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    lf = localforage_1.default.createInstance({
                        name: 'MagicAuthLocalStorageDB',
                        storeName: 'MagicAuthLocalStorage',
                    });
                    return [4 /*yield*/, lf.defineDriver(memoryDriver)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, lf.setDriver([localforage_1.default.INDEXEDDB, localforage_1.default.LOCALSTORAGE, memoryDriver._driver])];
                case 2:
                    _a.sent();
                    return [2 /*return*/, lf];
            }
        });
    }); },
});
