// @aws-sdk/client-sesv2/dist-es/commands/CreateContactCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateContactRequest, CreateContactResponse } from "../models/models_0";
import { deserializeAws_restJson1CreateContactCommand, serializeAws_restJson1CreateContactCommand, } from "../protocols/Aws_restJson1";
var CreateContactCommand = (function (_super) {
    __extends(CreateContactCommand, _super);
    function CreateContactCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateContactCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "CreateContactCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateContactRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateContactResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateContactCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1CreateContactCommand(input, context);
    };
    CreateContactCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1CreateContactCommand(output, context);
    };
    return CreateContactCommand;
}($Command));
export { CreateContactCommand };
