// @aws-sdk/client-sesv2/dist-es/commands/CreateCustomVerificationEmailTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateCustomVerificationEmailTemplateRequest, CreateCustomVerificationEmailTemplateResponse, } from "../models/models_0";
import { deserializeAws_restJson1CreateCustomVerificationEmailTemplateCommand, serializeAws_restJson1CreateCustomVerificationEmailTemplateCommand, } from "../protocols/Aws_restJson1";
var CreateCustomVerificationEmailTemplateCommand = (function (_super) {
    __extends(CreateCustomVerificationEmailTemplateCommand, _super);
    function CreateCustomVerificationEmailTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateCustomVerificationEmailTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "CreateCustomVerificationEmailTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateCustomVerificationEmailTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateCustomVerificationEmailTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateCustomVerificationEmailTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1CreateCustomVerificationEmailTemplateCommand(input, context);
    };
    CreateCustomVerificationEmailTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1CreateCustomVerificationEmailTemplateCommand(output, context);
    };
    return CreateCustomVerificationEmailTemplateCommand;
}($Command));
export { CreateCustomVerificationEmailTemplateCommand };
