// @aws-sdk/client-sesv2/dist-es/commands/CreateEmailIdentityCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateEmailIdentityRequest, CreateEmailIdentityResponse } from "../models/models_0";
import { deserializeAws_restJson1CreateEmailIdentityCommand, serializeAws_restJson1CreateEmailIdentityCommand, } from "../protocols/Aws_restJson1";
var CreateEmailIdentityCommand = (function (_super) {
    __extends(CreateEmailIdentityCommand, _super);
    function CreateEmailIdentityCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateEmailIdentityCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "CreateEmailIdentityCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateEmailIdentityRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CreateEmailIdentityResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateEmailIdentityCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1CreateEmailIdentityCommand(input, context);
    };
    CreateEmailIdentityCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1CreateEmailIdentityCommand(output, context);
    };
    return CreateEmailIdentityCommand;
}($Command));
export { CreateEmailIdentityCommand };
