// @aws-sdk/client-sesv2/dist-es/commands/DeleteDedicatedIpPoolCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteDedicatedIpPoolRequest, DeleteDedicatedIpPoolResponse } from "../models/models_0";
import { deserializeAws_restJson1DeleteDedicatedIpPoolCommand, serializeAws_restJson1DeleteDedicatedIpPoolCommand, } from "../protocols/Aws_restJson1";
var DeleteDedicatedIpPoolCommand = (function (_super) {
    __extends(DeleteDedicatedIpPoolCommand, _super);
    function DeleteDedicatedIpPoolCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteDedicatedIpPoolCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "DeleteDedicatedIpPoolCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteDedicatedIpPoolRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteDedicatedIpPoolResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteDedicatedIpPoolCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1DeleteDedicatedIpPoolCommand(input, context);
    };
    DeleteDedicatedIpPoolCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1DeleteDedicatedIpPoolCommand(output, context);
    };
    return DeleteDedicatedIpPoolCommand;
}($Command));
export { DeleteDedicatedIpPoolCommand };
