// @aws-sdk/client-sesv2/dist-es/commands/DeleteEmailIdentityPolicyCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteEmailIdentityPolicyRequest, DeleteEmailIdentityPolicyResponse } from "../models/models_0";
import { deserializeAws_restJson1DeleteEmailIdentityPolicyCommand, serializeAws_restJson1DeleteEmailIdentityPolicyCommand, } from "../protocols/Aws_restJson1";
var DeleteEmailIdentityPolicyCommand = (function (_super) {
    __extends(DeleteEmailIdentityPolicyCommand, _super);
    function DeleteEmailIdentityPolicyCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteEmailIdentityPolicyCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "DeleteEmailIdentityPolicyCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteEmailIdentityPolicyRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteEmailIdentityPolicyResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteEmailIdentityPolicyCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1DeleteEmailIdentityPolicyCommand(input, context);
    };
    DeleteEmailIdentityPolicyCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1DeleteEmailIdentityPolicyCommand(output, context);
    };
    return DeleteEmailIdentityPolicyCommand;
}($Command));
export { DeleteEmailIdentityPolicyCommand };
