// @aws-sdk/client-sesv2/dist-es/commands/DeleteCustomVerificationEmailTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteCustomVerificationEmailTemplateRequest, DeleteCustomVerificationEmailTemplateResponse, } from "../models/models_0";
import { deserializeAws_restJson1DeleteCustomVerificationEmailTemplateCommand, serializeAws_restJson1DeleteCustomVerificationEmailTemplateCommand, } from "../protocols/Aws_restJson1";
var DeleteCustomVerificationEmailTemplateCommand = (function (_super) {
    __extends(DeleteCustomVerificationEmailTemplateCommand, _super);
    function DeleteCustomVerificationEmailTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteCustomVerificationEmailTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "DeleteCustomVerificationEmailTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteCustomVerificationEmailTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteCustomVerificationEmailTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteCustomVerificationEmailTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1DeleteCustomVerificationEmailTemplateCommand(input, context);
    };
    DeleteCustomVerificationEmailTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1DeleteCustomVerificationEmailTemplateCommand(output, context);
    };
    return DeleteCustomVerificationEmailTemplateCommand;
}($Command));
export { DeleteCustomVerificationEmailTemplateCommand };
