// @aws-sdk/client-sesv2/dist-es/commands/GetAccountCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetAccountRequest, GetAccountResponse } from "../models/models_0";
import { deserializeAws_restJson1GetAccountCommand, serializeAws_restJson1GetAccountCommand, } from "../protocols/Aws_restJson1";
var GetAccountCommand = (function (_super) {
    __extends(GetAccountCommand, _super);
    function GetAccountCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetAccountCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetAccountCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetAccountRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetAccountResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetAccountCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetAccountCommand(input, context);
    };
    GetAccountCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetAccountCommand(output, context);
    };
    return GetAccountCommand;
}($Command));
export { GetAccountCommand };
