// @aws-sdk/client-sesv2/dist-es/commands/GetBlacklistReportsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetBlacklistReportsRequest, GetBlacklistReportsResponse } from "../models/models_0";
import { deserializeAws_restJson1GetBlacklistReportsCommand, serializeAws_restJson1GetBlacklistReportsCommand, } from "../protocols/Aws_restJson1";
var GetBlacklistReportsCommand = (function (_super) {
    __extends(GetBlacklistReportsCommand, _super);
    function GetBlacklistReportsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetBlacklistReportsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetBlacklistReportsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetBlacklistReportsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetBlacklistReportsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetBlacklistReportsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetBlacklistReportsCommand(input, context);
    };
    GetBlacklistReportsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetBlacklistReportsCommand(output, context);
    };
    return GetBlacklistReportsCommand;
}($Command));
export { GetBlacklistReportsCommand };
