// @aws-sdk/client-sesv2/dist-es/commands/GetCustomVerificationEmailTemplateCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetCustomVerificationEmailTemplateRequest, GetCustomVerificationEmailTemplateResponse, } from "../models/models_0";
import { deserializeAws_restJson1GetCustomVerificationEmailTemplateCommand, serializeAws_restJson1GetCustomVerificationEmailTemplateCommand, } from "../protocols/Aws_restJson1";
var GetCustomVerificationEmailTemplateCommand = (function (_super) {
    __extends(GetCustomVerificationEmailTemplateCommand, _super);
    function GetCustomVerificationEmailTemplateCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetCustomVerificationEmailTemplateCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetCustomVerificationEmailTemplateCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetCustomVerificationEmailTemplateRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetCustomVerificationEmailTemplateResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetCustomVerificationEmailTemplateCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetCustomVerificationEmailTemplateCommand(input, context);
    };
    GetCustomVerificationEmailTemplateCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetCustomVerificationEmailTemplateCommand(output, context);
    };
    return GetCustomVerificationEmailTemplateCommand;
}($Command));
export { GetCustomVerificationEmailTemplateCommand };
