// @aws-sdk/client-sesv2/dist-es/commands/GetContactListCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetContactListRequest, GetContactListResponse } from "../models/models_0";
import { deserializeAws_restJson1GetContactListCommand, serializeAws_restJson1GetContactListCommand, } from "../protocols/Aws_restJson1";
var GetContactListCommand = (function (_super) {
    __extends(GetContactListCommand, _super);
    function GetContactListCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetContactListCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetContactListCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetContactListRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetContactListResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetContactListCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetContactListCommand(input, context);
    };
    GetContactListCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetContactListCommand(output, context);
    };
    return GetContactListCommand;
}($Command));
export { GetContactListCommand };
