// @aws-sdk/client-sesv2/dist-es/commands/GetConfigurationSetCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetConfigurationSetRequest, GetConfigurationSetResponse } from "../models/models_0";
import { deserializeAws_restJson1GetConfigurationSetCommand, serializeAws_restJson1GetConfigurationSetCommand, } from "../protocols/Aws_restJson1";
var GetConfigurationSetCommand = (function (_super) {
    __extends(GetConfigurationSetCommand, _super);
    function GetConfigurationSetCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetConfigurationSetCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetConfigurationSetCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetConfigurationSetRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetConfigurationSetResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetConfigurationSetCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetConfigurationSetCommand(input, context);
    };
    GetConfigurationSetCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetConfigurationSetCommand(output, context);
    };
    return GetConfigurationSetCommand;
}($Command));
export { GetConfigurationSetCommand };
