// @aws-sdk/client-sesv2/dist-es/commands/GetDomainStatisticsReportCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetDomainStatisticsReportRequest, GetDomainStatisticsReportResponse } from "../models/models_0";
import { deserializeAws_restJson1GetDomainStatisticsReportCommand, serializeAws_restJson1GetDomainStatisticsReportCommand, } from "../protocols/Aws_restJson1";
var GetDomainStatisticsReportCommand = (function (_super) {
    __extends(GetDomainStatisticsReportCommand, _super);
    function GetDomainStatisticsReportCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetDomainStatisticsReportCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetDomainStatisticsReportCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetDomainStatisticsReportRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetDomainStatisticsReportResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetDomainStatisticsReportCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetDomainStatisticsReportCommand(input, context);
    };
    GetDomainStatisticsReportCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetDomainStatisticsReportCommand(output, context);
    };
    return GetDomainStatisticsReportCommand;
}($Command));
export { GetDomainStatisticsReportCommand };
