// @aws-sdk/client-sesv2/dist-es/commands/GetDedicatedIpCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetDedicatedIpRequest, GetDedicatedIpResponse } from "../models/models_0";
import { deserializeAws_restJson1GetDedicatedIpCommand, serializeAws_restJson1GetDedicatedIpCommand, } from "../protocols/Aws_restJson1";
var GetDedicatedIpCommand = (function (_super) {
    __extends(GetDedicatedIpCommand, _super);
    function GetDedicatedIpCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetDedicatedIpCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetDedicatedIpCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetDedicatedIpRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetDedicatedIpResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetDedicatedIpCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetDedicatedIpCommand(input, context);
    };
    GetDedicatedIpCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetDedicatedIpCommand(output, context);
    };
    return GetDedicatedIpCommand;
}($Command));
export { GetDedicatedIpCommand };
