// @aws-sdk/client-sesv2/dist-es/commands/GetDeliverabilityDashboardOptionsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetDeliverabilityDashboardOptionsRequest, GetDeliverabilityDashboardOptionsResponse, } from "../models/models_0";
import { deserializeAws_restJson1GetDeliverabilityDashboardOptionsCommand, serializeAws_restJson1GetDeliverabilityDashboardOptionsCommand, } from "../protocols/Aws_restJson1";
var GetDeliverabilityDashboardOptionsCommand = (function (_super) {
    __extends(GetDeliverabilityDashboardOptionsCommand, _super);
    function GetDeliverabilityDashboardOptionsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetDeliverabilityDashboardOptionsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetDeliverabilityDashboardOptionsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetDeliverabilityDashboardOptionsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetDeliverabilityDashboardOptionsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetDeliverabilityDashboardOptionsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetDeliverabilityDashboardOptionsCommand(input, context);
    };
    GetDeliverabilityDashboardOptionsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetDeliverabilityDashboardOptionsCommand(output, context);
    };
    return GetDeliverabilityDashboardOptionsCommand;
}($Command));
export { GetDeliverabilityDashboardOptionsCommand };
