// @aws-sdk/client-sesv2/dist-es/commands/GetDeliverabilityTestReportCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetDeliverabilityTestReportRequest, GetDeliverabilityTestReportResponse } from "../models/models_0";
import { deserializeAws_restJson1GetDeliverabilityTestReportCommand, serializeAws_restJson1GetDeliverabilityTestReportCommand, } from "../protocols/Aws_restJson1";
var GetDeliverabilityTestReportCommand = (function (_super) {
    __extends(GetDeliverabilityTestReportCommand, _super);
    function GetDeliverabilityTestReportCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetDeliverabilityTestReportCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetDeliverabilityTestReportCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetDeliverabilityTestReportRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetDeliverabilityTestReportResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetDeliverabilityTestReportCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetDeliverabilityTestReportCommand(input, context);
    };
    GetDeliverabilityTestReportCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetDeliverabilityTestReportCommand(output, context);
    };
    return GetDeliverabilityTestReportCommand;
}($Command));
export { GetDeliverabilityTestReportCommand };
