// @aws-sdk/client-sesv2/dist-es/commands/GetEmailIdentityPoliciesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetEmailIdentityPoliciesRequest, GetEmailIdentityPoliciesResponse } from "../models/models_0";
import { deserializeAws_restJson1GetEmailIdentityPoliciesCommand, serializeAws_restJson1GetEmailIdentityPoliciesCommand, } from "../protocols/Aws_restJson1";
var GetEmailIdentityPoliciesCommand = (function (_super) {
    __extends(GetEmailIdentityPoliciesCommand, _super);
    function GetEmailIdentityPoliciesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetEmailIdentityPoliciesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "GetEmailIdentityPoliciesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetEmailIdentityPoliciesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: GetEmailIdentityPoliciesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetEmailIdentityPoliciesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1GetEmailIdentityPoliciesCommand(input, context);
    };
    GetEmailIdentityPoliciesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1GetEmailIdentityPoliciesCommand(output, context);
    };
    return GetEmailIdentityPoliciesCommand;
}($Command));
export { GetEmailIdentityPoliciesCommand };
