// @aws-sdk/client-sesv2/dist-es/commands/ListDomainDeliverabilityCampaignsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListDomainDeliverabilityCampaignsRequest, ListDomainDeliverabilityCampaignsResponse, } from "../models/models_0";
import { deserializeAws_restJson1ListDomainDeliverabilityCampaignsCommand, serializeAws_restJson1ListDomainDeliverabilityCampaignsCommand, } from "../protocols/Aws_restJson1";
var ListDomainDeliverabilityCampaignsCommand = (function (_super) {
    __extends(ListDomainDeliverabilityCampaignsCommand, _super);
    function ListDomainDeliverabilityCampaignsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListDomainDeliverabilityCampaignsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListDomainDeliverabilityCampaignsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListDomainDeliverabilityCampaignsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListDomainDeliverabilityCampaignsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListDomainDeliverabilityCampaignsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListDomainDeliverabilityCampaignsCommand(input, context);
    };
    ListDomainDeliverabilityCampaignsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListDomainDeliverabilityCampaignsCommand(output, context);
    };
    return ListDomainDeliverabilityCampaignsCommand;
}($Command));
export { ListDomainDeliverabilityCampaignsCommand };
