// @aws-sdk/client-sesv2/dist-es/commands/ListDedicatedIpPoolsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListDedicatedIpPoolsRequest, ListDedicatedIpPoolsResponse } from "../models/models_0";
import { deserializeAws_restJson1ListDedicatedIpPoolsCommand, serializeAws_restJson1ListDedicatedIpPoolsCommand, } from "../protocols/Aws_restJson1";
var ListDedicatedIpPoolsCommand = (function (_super) {
    __extends(ListDedicatedIpPoolsCommand, _super);
    function ListDedicatedIpPoolsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListDedicatedIpPoolsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListDedicatedIpPoolsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListDedicatedIpPoolsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListDedicatedIpPoolsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListDedicatedIpPoolsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListDedicatedIpPoolsCommand(input, context);
    };
    ListDedicatedIpPoolsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListDedicatedIpPoolsCommand(output, context);
    };
    return ListDedicatedIpPoolsCommand;
}($Command));
export { ListDedicatedIpPoolsCommand };
