// @aws-sdk/client-sesv2/dist-es/commands/ListImportJobsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListImportJobsRequest, ListImportJobsResponse } from "../models/models_0";
import { deserializeAws_restJson1ListImportJobsCommand, serializeAws_restJson1ListImportJobsCommand, } from "../protocols/Aws_restJson1";
var ListImportJobsCommand = (function (_super) {
    __extends(ListImportJobsCommand, _super);
    function ListImportJobsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListImportJobsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListImportJobsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListImportJobsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListImportJobsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListImportJobsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListImportJobsCommand(input, context);
    };
    ListImportJobsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListImportJobsCommand(output, context);
    };
    return ListImportJobsCommand;
}($Command));
export { ListImportJobsCommand };
