// @aws-sdk/client-sesv2/dist-es/commands/ListConfigurationSetsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListConfigurationSetsRequest, ListConfigurationSetsResponse } from "../models/models_0";
import { deserializeAws_restJson1ListConfigurationSetsCommand, serializeAws_restJson1ListConfigurationSetsCommand, } from "../protocols/Aws_restJson1";
var ListConfigurationSetsCommand = (function (_super) {
    __extends(ListConfigurationSetsCommand, _super);
    function ListConfigurationSetsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListConfigurationSetsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListConfigurationSetsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListConfigurationSetsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListConfigurationSetsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListConfigurationSetsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListConfigurationSetsCommand(input, context);
    };
    ListConfigurationSetsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListConfigurationSetsCommand(output, context);
    };
    return ListConfigurationSetsCommand;
}($Command));
export { ListConfigurationSetsCommand };
