// @aws-sdk/client-sesv2/dist-es/commands/ListContactsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListContactsRequest, ListContactsResponse } from "../models/models_0";
import { deserializeAws_restJson1ListContactsCommand, serializeAws_restJson1ListContactsCommand, } from "../protocols/Aws_restJson1";
var ListContactsCommand = (function (_super) {
    __extends(ListContactsCommand, _super);
    function ListContactsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListContactsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListContactsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListContactsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListContactsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListContactsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListContactsCommand(input, context);
    };
    ListContactsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListContactsCommand(output, context);
    };
    return ListContactsCommand;
}($Command));
export { ListContactsCommand };
