// @aws-sdk/client-sesv2/dist-es/commands/ListEmailIdentitiesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListEmailIdentitiesRequest, ListEmailIdentitiesResponse } from "../models/models_0";
import { deserializeAws_restJson1ListEmailIdentitiesCommand, serializeAws_restJson1ListEmailIdentitiesCommand, } from "../protocols/Aws_restJson1";
var ListEmailIdentitiesCommand = (function (_super) {
    __extends(ListEmailIdentitiesCommand, _super);
    function ListEmailIdentitiesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListEmailIdentitiesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "ListEmailIdentitiesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListEmailIdentitiesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListEmailIdentitiesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListEmailIdentitiesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1ListEmailIdentitiesCommand(input, context);
    };
    ListEmailIdentitiesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1ListEmailIdentitiesCommand(output, context);
    };
    return ListEmailIdentitiesCommand;
}($Command));
export { ListEmailIdentitiesCommand };
