// @aws-sdk/client-sesv2/dist-es/commands/PutAccountDetailsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutAccountDetailsRequest, PutAccountDetailsResponse } from "../models/models_0";
import { deserializeAws_restJson1PutAccountDetailsCommand, serializeAws_restJson1PutAccountDetailsCommand, } from "../protocols/Aws_restJson1";
var PutAccountDetailsCommand = (function (_super) {
    __extends(PutAccountDetailsCommand, _super);
    function PutAccountDetailsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutAccountDetailsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutAccountDetailsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutAccountDetailsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutAccountDetailsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutAccountDetailsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutAccountDetailsCommand(input, context);
    };
    PutAccountDetailsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutAccountDetailsCommand(output, context);
    };
    return PutAccountDetailsCommand;
}($Command));
export { PutAccountDetailsCommand };
