// @aws-sdk/client-sesv2/dist-es/commands/PutAccountSendingAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutAccountSendingAttributesRequest, PutAccountSendingAttributesResponse } from "../models/models_0";
import { deserializeAws_restJson1PutAccountSendingAttributesCommand, serializeAws_restJson1PutAccountSendingAttributesCommand, } from "../protocols/Aws_restJson1";
var PutAccountSendingAttributesCommand = (function (_super) {
    __extends(PutAccountSendingAttributesCommand, _super);
    function PutAccountSendingAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutAccountSendingAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutAccountSendingAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutAccountSendingAttributesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutAccountSendingAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutAccountSendingAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutAccountSendingAttributesCommand(input, context);
    };
    PutAccountSendingAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutAccountSendingAttributesCommand(output, context);
    };
    return PutAccountSendingAttributesCommand;
}($Command));
export { PutAccountSendingAttributesCommand };
