// @aws-sdk/client-sesv2/dist-es/commands/PutConfigurationSetReputationOptionsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutConfigurationSetReputationOptionsRequest, PutConfigurationSetReputationOptionsResponse, } from "../models/models_0";
import { deserializeAws_restJson1PutConfigurationSetReputationOptionsCommand, serializeAws_restJson1PutConfigurationSetReputationOptionsCommand, } from "../protocols/Aws_restJson1";
var PutConfigurationSetReputationOptionsCommand = (function (_super) {
    __extends(PutConfigurationSetReputationOptionsCommand, _super);
    function PutConfigurationSetReputationOptionsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutConfigurationSetReputationOptionsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutConfigurationSetReputationOptionsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutConfigurationSetReputationOptionsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutConfigurationSetReputationOptionsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutConfigurationSetReputationOptionsCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutConfigurationSetReputationOptionsCommand(input, context);
    };
    PutConfigurationSetReputationOptionsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutConfigurationSetReputationOptionsCommand(output, context);
    };
    return PutConfigurationSetReputationOptionsCommand;
}($Command));
export { PutConfigurationSetReputationOptionsCommand };
