// @aws-sdk/client-sesv2/dist-es/commands/PutDedicatedIpInPoolCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutDedicatedIpInPoolRequest, PutDedicatedIpInPoolResponse } from "../models/models_0";
import { deserializeAws_restJson1PutDedicatedIpInPoolCommand, serializeAws_restJson1PutDedicatedIpInPoolCommand, } from "../protocols/Aws_restJson1";
var PutDedicatedIpInPoolCommand = (function (_super) {
    __extends(PutDedicatedIpInPoolCommand, _super);
    function PutDedicatedIpInPoolCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutDedicatedIpInPoolCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutDedicatedIpInPoolCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutDedicatedIpInPoolRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutDedicatedIpInPoolResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutDedicatedIpInPoolCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutDedicatedIpInPoolCommand(input, context);
    };
    PutDedicatedIpInPoolCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutDedicatedIpInPoolCommand(output, context);
    };
    return PutDedicatedIpInPoolCommand;
}($Command));
export { PutDedicatedIpInPoolCommand };
