// @aws-sdk/client-sesv2/dist-es/commands/PutDedicatedIpWarmupAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutDedicatedIpWarmupAttributesRequest, PutDedicatedIpWarmupAttributesResponse } from "../models/models_0";
import { deserializeAws_restJson1PutDedicatedIpWarmupAttributesCommand, serializeAws_restJson1PutDedicatedIpWarmupAttributesCommand, } from "../protocols/Aws_restJson1";
var PutDedicatedIpWarmupAttributesCommand = (function (_super) {
    __extends(PutDedicatedIpWarmupAttributesCommand, _super);
    function PutDedicatedIpWarmupAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutDedicatedIpWarmupAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutDedicatedIpWarmupAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutDedicatedIpWarmupAttributesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutDedicatedIpWarmupAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutDedicatedIpWarmupAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutDedicatedIpWarmupAttributesCommand(input, context);
    };
    PutDedicatedIpWarmupAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutDedicatedIpWarmupAttributesCommand(output, context);
    };
    return PutDedicatedIpWarmupAttributesCommand;
}($Command));
export { PutDedicatedIpWarmupAttributesCommand };
