// @aws-sdk/client-sesv2/dist-es/commands/PutEmailIdentityDkimAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutEmailIdentityDkimAttributesRequest, PutEmailIdentityDkimAttributesResponse } from "../models/models_0";
import { deserializeAws_restJson1PutEmailIdentityDkimAttributesCommand, serializeAws_restJson1PutEmailIdentityDkimAttributesCommand, } from "../protocols/Aws_restJson1";
var PutEmailIdentityDkimAttributesCommand = (function (_super) {
    __extends(PutEmailIdentityDkimAttributesCommand, _super);
    function PutEmailIdentityDkimAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutEmailIdentityDkimAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutEmailIdentityDkimAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutEmailIdentityDkimAttributesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutEmailIdentityDkimAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutEmailIdentityDkimAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutEmailIdentityDkimAttributesCommand(input, context);
    };
    PutEmailIdentityDkimAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutEmailIdentityDkimAttributesCommand(output, context);
    };
    return PutEmailIdentityDkimAttributesCommand;
}($Command));
export { PutEmailIdentityDkimAttributesCommand };
