// @aws-sdk/client-sesv2/dist-es/commands/PutEmailIdentityFeedbackAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutEmailIdentityFeedbackAttributesRequest, PutEmailIdentityFeedbackAttributesResponse, } from "../models/models_0";
import { deserializeAws_restJson1PutEmailIdentityFeedbackAttributesCommand, serializeAws_restJson1PutEmailIdentityFeedbackAttributesCommand, } from "../protocols/Aws_restJson1";
var PutEmailIdentityFeedbackAttributesCommand = (function (_super) {
    __extends(PutEmailIdentityFeedbackAttributesCommand, _super);
    function PutEmailIdentityFeedbackAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutEmailIdentityFeedbackAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "PutEmailIdentityFeedbackAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutEmailIdentityFeedbackAttributesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutEmailIdentityFeedbackAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutEmailIdentityFeedbackAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1PutEmailIdentityFeedbackAttributesCommand(input, context);
    };
    PutEmailIdentityFeedbackAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1PutEmailIdentityFeedbackAttributesCommand(output, context);
    };
    return PutEmailIdentityFeedbackAttributesCommand;
}($Command));
export { PutEmailIdentityFeedbackAttributesCommand };
