// @aws-sdk/client-sesv2/dist-es/commands/SendBulkEmailCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendBulkEmailRequest, SendBulkEmailResponse } from "../models/models_0";
import { deserializeAws_restJson1SendBulkEmailCommand, serializeAws_restJson1SendBulkEmailCommand, } from "../protocols/Aws_restJson1";
var SendBulkEmailCommand = (function (_super) {
    __extends(SendBulkEmailCommand, _super);
    function SendBulkEmailCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SendBulkEmailCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "SendBulkEmailCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SendBulkEmailRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SendBulkEmailResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SendBulkEmailCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1SendBulkEmailCommand(input, context);
    };
    SendBulkEmailCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1SendBulkEmailCommand(output, context);
    };
    return SendBulkEmailCommand;
}($Command));
export { SendBulkEmailCommand };
