// @aws-sdk/client-sesv2/dist-es/commands/SendEmailCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SendEmailRequest, SendEmailResponse } from "../models/models_0";
import { deserializeAws_restJson1SendEmailCommand, serializeAws_restJson1SendEmailCommand, } from "../protocols/Aws_restJson1";
var SendEmailCommand = (function (_super) {
    __extends(SendEmailCommand, _super);
    function SendEmailCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SendEmailCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "SendEmailCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SendEmailRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SendEmailResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SendEmailCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1SendEmailCommand(input, context);
    };
    SendEmailCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1SendEmailCommand(output, context);
    };
    return SendEmailCommand;
}($Command));
export { SendEmailCommand };
