// @aws-sdk/client-sesv2/dist-es/commands/UpdateContactCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { UpdateContactRequest, UpdateContactResponse } from "../models/models_0";
import { deserializeAws_restJson1UpdateContactCommand, serializeAws_restJson1UpdateContactCommand, } from "../protocols/Aws_restJson1";
var UpdateContactCommand = (function (_super) {
    __extends(UpdateContactCommand, _super);
    function UpdateContactCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    UpdateContactCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESv2Client";
        var commandName = "UpdateContactCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: UpdateContactRequest.filterSensitiveLog,
            outputFilterSensitiveLog: UpdateContactResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    UpdateContactCommand.prototype.serialize = function (input, context) {
        return serializeAws_restJson1UpdateContactCommand(input, context);
    };
    UpdateContactCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_restJson1UpdateContactCommand(output, context);
    };
    return UpdateContactCommand;
}($Command));
export { UpdateContactCommand };
