// @aws-sdk/client-sns/dist-es/commands/CheckIfPhoneNumberIsOptedOutCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CheckIfPhoneNumberIsOptedOutInput, CheckIfPhoneNumberIsOptedOutResponse } from "../models/models_0";
import { deserializeAws_queryCheckIfPhoneNumberIsOptedOutCommand, serializeAws_queryCheckIfPhoneNumberIsOptedOutCommand, } from "../protocols/Aws_query";
var CheckIfPhoneNumberIsOptedOutCommand = (function (_super) {
    __extends(CheckIfPhoneNumberIsOptedOutCommand, _super);
    function CheckIfPhoneNumberIsOptedOutCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CheckIfPhoneNumberIsOptedOutCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "CheckIfPhoneNumberIsOptedOutCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CheckIfPhoneNumberIsOptedOutInput.filterSensitiveLog,
            outputFilterSensitiveLog: CheckIfPhoneNumberIsOptedOutResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CheckIfPhoneNumberIsOptedOutCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCheckIfPhoneNumberIsOptedOutCommand(input, context);
    };
    CheckIfPhoneNumberIsOptedOutCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCheckIfPhoneNumberIsOptedOutCommand(output, context);
    };
    return CheckIfPhoneNumberIsOptedOutCommand;
}($Command));
export { CheckIfPhoneNumberIsOptedOutCommand };
