// @aws-sdk/client-sns/dist-es/commands/CreatePlatformEndpointCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateEndpointResponse, CreatePlatformEndpointInput } from "../models/models_0";
import { deserializeAws_queryCreatePlatformEndpointCommand, serializeAws_queryCreatePlatformEndpointCommand, } from "../protocols/Aws_query";
var CreatePlatformEndpointCommand = (function (_super) {
    __extends(CreatePlatformEndpointCommand, _super);
    function CreatePlatformEndpointCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreatePlatformEndpointCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "CreatePlatformEndpointCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreatePlatformEndpointInput.filterSensitiveLog,
            outputFilterSensitiveLog: CreateEndpointResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreatePlatformEndpointCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreatePlatformEndpointCommand(input, context);
    };
    CreatePlatformEndpointCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreatePlatformEndpointCommand(output, context);
    };
    return CreatePlatformEndpointCommand;
}($Command));
export { CreatePlatformEndpointCommand };
