// @aws-sdk/client-sns/dist-es/commands/CreateSMSSandboxPhoneNumberCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CreateSMSSandboxPhoneNumberInput, CreateSMSSandboxPhoneNumberResult } from "../models/models_0";
import { deserializeAws_queryCreateSMSSandboxPhoneNumberCommand, serializeAws_queryCreateSMSSandboxPhoneNumberCommand, } from "../protocols/Aws_query";
var CreateSMSSandboxPhoneNumberCommand = (function (_super) {
    __extends(CreateSMSSandboxPhoneNumberCommand, _super);
    function CreateSMSSandboxPhoneNumberCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CreateSMSSandboxPhoneNumberCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "CreateSMSSandboxPhoneNumberCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CreateSMSSandboxPhoneNumberInput.filterSensitiveLog,
            outputFilterSensitiveLog: CreateSMSSandboxPhoneNumberResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CreateSMSSandboxPhoneNumberCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryCreateSMSSandboxPhoneNumberCommand(input, context);
    };
    CreateSMSSandboxPhoneNumberCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryCreateSMSSandboxPhoneNumberCommand(output, context);
    };
    return CreateSMSSandboxPhoneNumberCommand;
}($Command));
export { CreateSMSSandboxPhoneNumberCommand };
