// @aws-sdk/client-sns/dist-es/commands/DeleteSMSSandboxPhoneNumberCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteSMSSandboxPhoneNumberInput, DeleteSMSSandboxPhoneNumberResult } from "../models/models_0";
import { deserializeAws_queryDeleteSMSSandboxPhoneNumberCommand, serializeAws_queryDeleteSMSSandboxPhoneNumberCommand, } from "../protocols/Aws_query";
var DeleteSMSSandboxPhoneNumberCommand = (function (_super) {
    __extends(DeleteSMSSandboxPhoneNumberCommand, _super);
    function DeleteSMSSandboxPhoneNumberCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteSMSSandboxPhoneNumberCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "DeleteSMSSandboxPhoneNumberCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteSMSSandboxPhoneNumberInput.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteSMSSandboxPhoneNumberResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteSMSSandboxPhoneNumberCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryDeleteSMSSandboxPhoneNumberCommand(input, context);
    };
    DeleteSMSSandboxPhoneNumberCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryDeleteSMSSandboxPhoneNumberCommand(output, context);
    };
    return DeleteSMSSandboxPhoneNumberCommand;
}($Command));
export { DeleteSMSSandboxPhoneNumberCommand };
