// @aws-sdk/client-sns/dist-es/commands/GetSubscriptionAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetSubscriptionAttributesInput, GetSubscriptionAttributesResponse } from "../models/models_0";
import { deserializeAws_queryGetSubscriptionAttributesCommand, serializeAws_queryGetSubscriptionAttributesCommand, } from "../protocols/Aws_query";
var GetSubscriptionAttributesCommand = (function (_super) {
    __extends(GetSubscriptionAttributesCommand, _super);
    function GetSubscriptionAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    GetSubscriptionAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "GetSubscriptionAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: GetSubscriptionAttributesInput.filterSensitiveLog,
            outputFilterSensitiveLog: GetSubscriptionAttributesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    GetSubscriptionAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryGetSubscriptionAttributesCommand(input, context);
    };
    GetSubscriptionAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryGetSubscriptionAttributesCommand(output, context);
    };
    return GetSubscriptionAttributesCommand;
}($Command));
export { GetSubscriptionAttributesCommand };
