// @aws-sdk/client-sns/dist-es/commands/ListPlatformApplicationsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListPlatformApplicationsInput, ListPlatformApplicationsResponse } from "../models/models_0";
import { deserializeAws_queryListPlatformApplicationsCommand, serializeAws_queryListPlatformApplicationsCommand, } from "../protocols/Aws_query";
var ListPlatformApplicationsCommand = (function (_super) {
    __extends(ListPlatformApplicationsCommand, _super);
    function ListPlatformApplicationsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListPlatformApplicationsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "ListPlatformApplicationsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListPlatformApplicationsInput.filterSensitiveLog,
            outputFilterSensitiveLog: ListPlatformApplicationsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListPlatformApplicationsCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryListPlatformApplicationsCommand(input, context);
    };
    ListPlatformApplicationsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryListPlatformApplicationsCommand(output, context);
    };
    return ListPlatformApplicationsCommand;
}($Command));
export { ListPlatformApplicationsCommand };
