// @aws-sdk/client-sns/dist-es/commands/ListSMSSandboxPhoneNumbersCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListSMSSandboxPhoneNumbersInput, ListSMSSandboxPhoneNumbersResult } from "../models/models_0";
import { deserializeAws_queryListSMSSandboxPhoneNumbersCommand, serializeAws_queryListSMSSandboxPhoneNumbersCommand, } from "../protocols/Aws_query";
var ListSMSSandboxPhoneNumbersCommand = (function (_super) {
    __extends(ListSMSSandboxPhoneNumbersCommand, _super);
    function ListSMSSandboxPhoneNumbersCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListSMSSandboxPhoneNumbersCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "ListSMSSandboxPhoneNumbersCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListSMSSandboxPhoneNumbersInput.filterSensitiveLog,
            outputFilterSensitiveLog: ListSMSSandboxPhoneNumbersResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListSMSSandboxPhoneNumbersCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryListSMSSandboxPhoneNumbersCommand(input, context);
    };
    ListSMSSandboxPhoneNumbersCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryListSMSSandboxPhoneNumbersCommand(output, context);
    };
    return ListSMSSandboxPhoneNumbersCommand;
}($Command));
export { ListSMSSandboxPhoneNumbersCommand };
