// @aws-sdk/client-sns/dist-es/commands/PublishBatchCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PublishBatchInput, PublishBatchResponse } from "../models/models_0";
import { deserializeAws_queryPublishBatchCommand, serializeAws_queryPublishBatchCommand } from "../protocols/Aws_query";
var PublishBatchCommand = (function (_super) {
    __extends(PublishBatchCommand, _super);
    function PublishBatchCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PublishBatchCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "PublishBatchCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PublishBatchInput.filterSensitiveLog,
            outputFilterSensitiveLog: PublishBatchResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PublishBatchCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryPublishBatchCommand(input, context);
    };
    PublishBatchCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryPublishBatchCommand(output, context);
    };
    return PublishBatchCommand;
}($Command));
export { PublishBatchCommand };
