// @aws-sdk/client-sns/dist-es/commands/SetPlatformApplicationAttributesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { SetPlatformApplicationAttributesInput } from "../models/models_0";
import { deserializeAws_querySetPlatformApplicationAttributesCommand, serializeAws_querySetPlatformApplicationAttributesCommand, } from "../protocols/Aws_query";
var SetPlatformApplicationAttributesCommand = (function (_super) {
    __extends(SetPlatformApplicationAttributesCommand, _super);
    function SetPlatformApplicationAttributesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    SetPlatformApplicationAttributesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "SetPlatformApplicationAttributesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SetPlatformApplicationAttributesInput.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SetPlatformApplicationAttributesCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySetPlatformApplicationAttributesCommand(input, context);
    };
    SetPlatformApplicationAttributesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySetPlatformApplicationAttributesCommand(output, context);
    };
    return SetPlatformApplicationAttributesCommand;
}($Command));
export { SetPlatformApplicationAttributesCommand };
