// @aws-sdk/client-sns/dist-es/commands/VerifySMSSandboxPhoneNumberCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { VerifySMSSandboxPhoneNumberInput, VerifySMSSandboxPhoneNumberResult } from "../models/models_0";
import { deserializeAws_queryVerifySMSSandboxPhoneNumberCommand, serializeAws_queryVerifySMSSandboxPhoneNumberCommand, } from "../protocols/Aws_query";
var VerifySMSSandboxPhoneNumberCommand = (function (_super) {
    __extends(VerifySMSSandboxPhoneNumberCommand, _super);
    function VerifySMSSandboxPhoneNumberCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    VerifySMSSandboxPhoneNumberCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SNSClient";
        var commandName = "VerifySMSSandboxPhoneNumberCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: VerifySMSSandboxPhoneNumberInput.filterSensitiveLog,
            outputFilterSensitiveLog: VerifySMSSandboxPhoneNumberResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    VerifySMSSandboxPhoneNumberCommand.prototype.serialize = function (input, context) {
        return serializeAws_queryVerifySMSSandboxPhoneNumberCommand(input, context);
    };
    VerifySMSSandboxPhoneNumberCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_queryVerifySMSSandboxPhoneNumberCommand(output, context);
    };
    return VerifySMSSandboxPhoneNumberCommand;
}($Command));
export { VerifySMSSandboxPhoneNumberCommand };
