// @magic-sdk/provider/dist/cjs/util/storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sdk_environment_1 = require("../core/sdk-environment");
var lf;
/**
 * Proxies `localforage` methods with strong-typing.
 */
function proxyLocalForageMethod(method) {
    var _this = this;
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!lf) return [3 /*break*/, 2];
                        return [4 /*yield*/, sdk_environment_1.SDKEnvironment.configureStorage()];
                    case 1:
                        lf = _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, lf.ready()];
                    case 3:
                        _a.sent();
                        return [2 /*return*/, lf[method].apply(lf, tslib_1.__spread(args))];
                }
            });
        });
    };
}
exports.getItem = proxyLocalForageMethod('getItem');
exports.setItem = proxyLocalForageMethod('setItem');
exports.removeItem = proxyLocalForageMethod('removeItem');
exports.clear = proxyLocalForageMethod('clear');
exports.length = proxyLocalForageMethod('length');
exports.key = proxyLocalForageMethod('key');
exports.keys = proxyLocalForageMethod('keys');
exports.iterate = proxyLocalForageMethod('iterate');
