// @aws-sdk/client-sns/dist-cjs/SNS.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SNS = void 0;
const AddPermissionCommand_1 = require("./commands/AddPermissionCommand");
const CheckIfPhoneNumberIsOptedOutCommand_1 = require("./commands/CheckIfPhoneNumberIsOptedOutCommand");
const ConfirmSubscriptionCommand_1 = require("./commands/ConfirmSubscriptionCommand");
const CreatePlatformApplicationCommand_1 = require("./commands/CreatePlatformApplicationCommand");
const CreatePlatformEndpointCommand_1 = require("./commands/CreatePlatformEndpointCommand");
const CreateSMSSandboxPhoneNumberCommand_1 = require("./commands/CreateSMSSandboxPhoneNumberCommand");
const CreateTopicCommand_1 = require("./commands/CreateTopicCommand");
const DeleteEndpointCommand_1 = require("./commands/DeleteEndpointCommand");
const DeletePlatformApplicationCommand_1 = require("./commands/DeletePlatformApplicationCommand");
const DeleteSMSSandboxPhoneNumberCommand_1 = require("./commands/DeleteSMSSandboxPhoneNumberCommand");
const DeleteTopicCommand_1 = require("./commands/DeleteTopicCommand");
const GetEndpointAttributesCommand_1 = require("./commands/GetEndpointAttributesCommand");
const GetPlatformApplicationAttributesCommand_1 = require("./commands/GetPlatformApplicationAttributesCommand");
const GetSMSAttributesCommand_1 = require("./commands/GetSMSAttributesCommand");
const GetSMSSandboxAccountStatusCommand_1 = require("./commands/GetSMSSandboxAccountStatusCommand");
const GetSubscriptionAttributesCommand_1 = require("./commands/GetSubscriptionAttributesCommand");
const GetTopicAttributesCommand_1 = require("./commands/GetTopicAttributesCommand");
const ListEndpointsByPlatformApplicationCommand_1 = require("./commands/ListEndpointsByPlatformApplicationCommand");
const ListOriginationNumbersCommand_1 = require("./commands/ListOriginationNumbersCommand");
const ListPhoneNumbersOptedOutCommand_1 = require("./commands/ListPhoneNumbersOptedOutCommand");
const ListPlatformApplicationsCommand_1 = require("./commands/ListPlatformApplicationsCommand");
const ListSMSSandboxPhoneNumbersCommand_1 = require("./commands/ListSMSSandboxPhoneNumbersCommand");
const ListSubscriptionsByTopicCommand_1 = require("./commands/ListSubscriptionsByTopicCommand");
const ListSubscriptionsCommand_1 = require("./commands/ListSubscriptionsCommand");
const ListTagsForResourceCommand_1 = require("./commands/ListTagsForResourceCommand");
const ListTopicsCommand_1 = require("./commands/ListTopicsCommand");
const OptInPhoneNumberCommand_1 = require("./commands/OptInPhoneNumberCommand");
const PublishBatchCommand_1 = require("./commands/PublishBatchCommand");
const PublishCommand_1 = require("./commands/PublishCommand");
const RemovePermissionCommand_1 = require("./commands/RemovePermissionCommand");
const SetEndpointAttributesCommand_1 = require("./commands/SetEndpointAttributesCommand");
const SetPlatformApplicationAttributesCommand_1 = require("./commands/SetPlatformApplicationAttributesCommand");
const SetSMSAttributesCommand_1 = require("./commands/SetSMSAttributesCommand");
const SetSubscriptionAttributesCommand_1 = require("./commands/SetSubscriptionAttributesCommand");
const SetTopicAttributesCommand_1 = require("./commands/SetTopicAttributesCommand");
const SubscribeCommand_1 = require("./commands/SubscribeCommand");
const TagResourceCommand_1 = require("./commands/TagResourceCommand");
const UnsubscribeCommand_1 = require("./commands/UnsubscribeCommand");
const UntagResourceCommand_1 = require("./commands/UntagResourceCommand");
const VerifySMSSandboxPhoneNumberCommand_1 = require("./commands/VerifySMSSandboxPhoneNumberCommand");
const SNSClient_1 = require("./SNSClient");
class SNS extends SNSClient_1.SNSClient {
    addPermission(args, optionsOrCb, cb) {
        const command = new AddPermissionCommand_1.AddPermissionCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    checkIfPhoneNumberIsOptedOut(args, optionsOrCb, cb) {
        const command = new CheckIfPhoneNumberIsOptedOutCommand_1.CheckIfPhoneNumberIsOptedOutCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    confirmSubscription(args, optionsOrCb, cb) {
        const command = new ConfirmSubscriptionCommand_1.ConfirmSubscriptionCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    createPlatformApplication(args, optionsOrCb, cb) {
        const command = new CreatePlatformApplicationCommand_1.CreatePlatformApplicationCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    createPlatformEndpoint(args, optionsOrCb, cb) {
        const command = new CreatePlatformEndpointCommand_1.CreatePlatformEndpointCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    createSMSSandboxPhoneNumber(args, optionsOrCb, cb) {
        const command = new CreateSMSSandboxPhoneNumberCommand_1.CreateSMSSandboxPhoneNumberCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    createTopic(args, optionsOrCb, cb) {
        const command = new CreateTopicCommand_1.CreateTopicCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    deleteEndpoint(args, optionsOrCb, cb) {
        const command = new DeleteEndpointCommand_1.DeleteEndpointCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    deletePlatformApplication(args, optionsOrCb, cb) {
        const command = new DeletePlatformApplicationCommand_1.DeletePlatformApplicationCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    deleteSMSSandboxPhoneNumber(args, optionsOrCb, cb) {
        const command = new DeleteSMSSandboxPhoneNumberCommand_1.DeleteSMSSandboxPhoneNumberCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    deleteTopic(args, optionsOrCb, cb) {
        const command = new DeleteTopicCommand_1.DeleteTopicCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getEndpointAttributes(args, optionsOrCb, cb) {
        const command = new GetEndpointAttributesCommand_1.GetEndpointAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getPlatformApplicationAttributes(args, optionsOrCb, cb) {
        const command = new GetPlatformApplicationAttributesCommand_1.GetPlatformApplicationAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getSMSAttributes(args, optionsOrCb, cb) {
        const command = new GetSMSAttributesCommand_1.GetSMSAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getSMSSandboxAccountStatus(args, optionsOrCb, cb) {
        const command = new GetSMSSandboxAccountStatusCommand_1.GetSMSSandboxAccountStatusCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getSubscriptionAttributes(args, optionsOrCb, cb) {
        const command = new GetSubscriptionAttributesCommand_1.GetSubscriptionAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    getTopicAttributes(args, optionsOrCb, cb) {
        const command = new GetTopicAttributesCommand_1.GetTopicAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listEndpointsByPlatformApplication(args, optionsOrCb, cb) {
        const command = new ListEndpointsByPlatformApplicationCommand_1.ListEndpointsByPlatformApplicationCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listOriginationNumbers(args, optionsOrCb, cb) {
        const command = new ListOriginationNumbersCommand_1.ListOriginationNumbersCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listPhoneNumbersOptedOut(args, optionsOrCb, cb) {
        const command = new ListPhoneNumbersOptedOutCommand_1.ListPhoneNumbersOptedOutCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listPlatformApplications(args, optionsOrCb, cb) {
        const command = new ListPlatformApplicationsCommand_1.ListPlatformApplicationsCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listSMSSandboxPhoneNumbers(args, optionsOrCb, cb) {
        const command = new ListSMSSandboxPhoneNumbersCommand_1.ListSMSSandboxPhoneNumbersCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listSubscriptions(args, optionsOrCb, cb) {
        const command = new ListSubscriptionsCommand_1.ListSubscriptionsCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listSubscriptionsByTopic(args, optionsOrCb, cb) {
        const command = new ListSubscriptionsByTopicCommand_1.ListSubscriptionsByTopicCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listTagsForResource(args, optionsOrCb, cb) {
        const command = new ListTagsForResourceCommand_1.ListTagsForResourceCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    listTopics(args, optionsOrCb, cb) {
        const command = new ListTopicsCommand_1.ListTopicsCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    optInPhoneNumber(args, optionsOrCb, cb) {
        const command = new OptInPhoneNumberCommand_1.OptInPhoneNumberCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    publish(args, optionsOrCb, cb) {
        const command = new PublishCommand_1.PublishCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    publishBatch(args, optionsOrCb, cb) {
        const command = new PublishBatchCommand_1.PublishBatchCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    removePermission(args, optionsOrCb, cb) {
        const command = new RemovePermissionCommand_1.RemovePermissionCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    setEndpointAttributes(args, optionsOrCb, cb) {
        const command = new SetEndpointAttributesCommand_1.SetEndpointAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    setPlatformApplicationAttributes(args, optionsOrCb, cb) {
        const command = new SetPlatformApplicationAttributesCommand_1.SetPlatformApplicationAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    setSMSAttributes(args, optionsOrCb, cb) {
        const command = new SetSMSAttributesCommand_1.SetSMSAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    setSubscriptionAttributes(args, optionsOrCb, cb) {
        const command = new SetSubscriptionAttributesCommand_1.SetSubscriptionAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    setTopicAttributes(args, optionsOrCb, cb) {
        const command = new SetTopicAttributesCommand_1.SetTopicAttributesCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    subscribe(args, optionsOrCb, cb) {
        const command = new SubscribeCommand_1.SubscribeCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    tagResource(args, optionsOrCb, cb) {
        const command = new TagResourceCommand_1.TagResourceCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    unsubscribe(args, optionsOrCb, cb) {
        const command = new UnsubscribeCommand_1.UnsubscribeCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    untagResource(args, optionsOrCb, cb) {
        const command = new UntagResourceCommand_1.UntagResourceCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
    verifySMSSandboxPhoneNumber(args, optionsOrCb, cb) {
        const command = new VerifySMSSandboxPhoneNumberCommand_1.VerifySMSSandboxPhoneNumberCommand(args);
        if (typeof optionsOrCb === "function") {
            this.send(command, optionsOrCb);
        }
        else if (typeof cb === "function") {
            if (typeof optionsOrCb !== "object")
                throw new Error(`Expect http options but get ${typeof optionsOrCb}`);
            this.send(command, optionsOrCb || {}, cb);
        }
        else {
            return this.send(command, optionsOrCb);
        }
    }
}
exports.SNS = SNS;
